### PR TITLE
Capstone Project [WIP]

### PR DESCRIPTION
**Pull request by s473573**
A steganography command-line program to hide your deepest secrets in your favorite le sadfrog memes. _Very_ under-construction for now.

Examples:

`stool insert  _image.jpg_  _out-image.jpg_ "secret"`

`stool extract img.png`
`stool extract -k "mypass" safe-secret.png`

A file named `secret` will be created with your plaintext secret for everyone to point at.

### CODE REVIEW

**Review by VictoriaGrasshopper — APPROVED** [review]

**Review by vsmysle — APPROVED** [review]
Overall, the implementation is quite impressive! 👍  

I like the organization of the code and find the project idea intriguing. Moreover, I particularly admire the utilization of the `ring` library and `CHACHA20_POLY1305` encryption algorithm. If you aim to develop something larger and more practical, remember to transition from `ring::aead::UnboundKey` to `ring::aead::SealingKey` or `ring::aead::OpeningKey`.

**Comment by NamesMark** [comment]
I liked that you utilised clap, thiserror and tracing. Good approach to cli tools!

Very thorough documentation, kudos!

Also very cool that you described more than one method :)

**Comment by NamesMark** [comment]
Minor nitpick: you provide this usage example: `cargo run -- -k "mypass" insert image.png output.png "literally a secret"`, while your arguments are these:
```
Commands:
  insert   Makes a secret
  extract  Extracts a secret from a file, not always a meaningful one
  help     Print this message or the help of the given subcommand(s)

Options:
  -m, --method <METHOD>      [default: least_bit]
  -p, --password <PASSWORD>  Apply a password to secure your secret with symmetrical encryption
  -h, --help                 Print help
  -V, --version              Print version
  ```
  
  I suspect you changed the arg from key to password at some point :)

File: src/util/crypt.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    let nonce = aead::Nonce::assume_unique_for_key(nonce);
+    debug!("got nonce");
+
+    let mut ciphertext = message_data[12..].to_owned();
+    debug!("got ciphertext");
+    let plaintext = key.open_in_place(nonce, aead::Aad::empty(), &mut ciphertext)?;
+    debug!("successfully decoded");
+
+    Ok(plaintext.to_vec())
+}
```
Comment (NamesMark) [inline src/util/crypt.rs:50]: What's the maximum message length that's allowed here? I tested on a bit longer strings, it encrypts them seemingly fine, but crashes in decrypt() on key.open_in_place() stage.

File: src/steg/mod.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum SteganographyMethod {
+    // Embeds a secret by altering the frequency components of a JPEG image, very imperceptible
+    DCTC,
+    // Hides data withing the least significant bits of pixel values
+    LeastBit,
+    /// Annexes a file to the end of the image, barely a steganography but still useful for hiding archives
+    AttachZip,
+}
```
Comment (NamesMark) [inline src/steg/mod.rs:44]: Just curious: are you thinking to also implement DCTC later?

Comment (NamesMark) [reply src/steg/mod.rs:44]: (you don't have to do anything, just wondering)

**Comment by s473573** [comment]
yooo thanks for the warm feedback guys! <3

**Review by StaraShkapa — APPROVED** [review]
You could give user a choice of output directory, encryption algorithm is well implemented.

**Review by NamesMark — APPROVED** [review]

**Comment by b1ackych** [comment]
Overall the code is really well designed and lit!! The only thing I could advise is to encode any sequence of bytes, not just String.

I find the idea of steganography very intriguing and unique, hiding information to an extent that (almost) no one, apart from the sender and intended recipient will ever find out it existed.

**Review by yannyv2 — APPROVED** [review]
looks good to me :) 👍

**Review by bugsweeper — APPROVED** [review]
Looks good to me

**Pull request merged**